### ISSUE TITLE
Add ability to crop a MediaStream obtained through the share-this-tab API

### ISSUE THREAD

**Issue by eladalon1983** [w3c/mediacapture-screen-share#158]
Assume a MediaStream is obtained through the [share-this-tab API](https://github.com/w3c/mediacapture-screen-share/pull/148) (which is still under discussion). The application in question might wish to crop the video before sharing it remote; for example remove a portion of the viewport where it displays private information. One possible use case is an application like Google Slides streaming itself to a Google Meet session, after cropping away speaker notes.

However, user-browser interaction such as window-resizing, scrolling or zooming, can move DOM elements in the viewport. This happens asynchronously from the JS event loop. An application holding a frame, does not know whether said frame was captured before or after a resize/scroll/zoom event, and therefore **cannot crop robustly**.

Some non-solutions were briefly explored in [January's WG Slides](https://docs.google.com/presentation/d/1crumgYj4eHkjo04faLktPTg0QoYJhTFoosEBudfJBuw/edit#slide=id.g7954c29f8a_2_0), leading to the conclusion that an API addition/modification is necessary in order to support cropping functionality. An initial draft of a solution was presented in the slides. We did not finish the discussion during the meeting. I suggest we continue it here.

**Comment by eladalon1983** [comment]
Quoting @jan-ivar from a [related thread](https://github.com/w3c/mediacapture-screen-share/issues/156) and responding inline:

> @eladalon1983 Since we ran out of time in the meeting I'm adding some feedback here. I'm going to ignore that we disagree on the security properties of the proposed API for a moment, to see if we can make progress on cropping orthogonally to it:

Agreed - this is orthogonal. I will not misconstrue this discussion as implicit agreement over share-this-tab.

> > From [slides](https://docs.google.com/presentation/d/1crumgYj4eHkjo04faLktPTg0QoYJhTFoosEBudfJBuw/edit#slide=id.g7954c29f8a_2_14)
> > <img alt="" width="250" src="https://user-images.githubusercontent.com/3136226/105072152-84b26000-5a53-11eb-89dc-b2805846624a.png">
> 
> If I understand correctly, a top-level page from site A is capturing its tab and wants to crop the output to only include a sub-region _inside_ a sub-iframe containing a participating site B, Is that right? And cropping needs to be robust and frame-accurate enough to not accidentally share even a split-frame more, to avoid revealing sensitive info like speaker notes?

Sorry, that's not the setup here. Instead of sites A/B I'll use sites L/R for left/right to make it easier to follow. The setup discussed has site L as its top-level (Slides), and it embeds site R (Meet), which is on the right. Site R is an application centered on video conferencing, and contains significant logic to handle them; site L does not have any video-conferencing logic. Both sites are collaborating, so they can postMessage anything that's serializable (e.g. IDs). Site L will communicate to its collaborator (R) that "the div you're looking to share is <crop_id>." Site R will then instruct **the browser** to crop to that crop_id - whatever it might represent.

> If so, then this seems problematic on a couple of levels. For one, the iframe may be in a separate process, which might slip up on a frame here and there. And inherently having the party that decides what area should be cropped separate from the source being cropped seems like a dependency that would be hard to manage e.g. requiring different teams to coordinate (no pun intended).

Although sites L and R are running in different processes, such coordination is possible for the browser - which is ultimately in charge of stitching together all frames in its compositing stage. It would not have been possible/feasible for anything other than the browser, I tried to argue in the slides (I'll also produce an explainer).

> If we don't use explicit coordinates (which I could easily see going out of sync), relying on some agreed-upon `<div>` would also mean adding code to fish out said div from the iframe, and things can go wrong there a well (site B updating the `'id`). Also, site A would not see the `<div>` in site B's page unless they were same orgin.

I propose a new type of ID (in [my last slide](https://docs.google.com/presentation/d/1crumgYj4eHkjo04faLktPTg0QoYJhTFoosEBudfJBuw/edit#slide=id.g7954c29f8a_2_44) - the one with the code sample - I called it "crop_id") that will be used explicitly for cropping. This will prevent breakage over current applications if they modify their IDs during runtime. A well-behaved application will not change the crop-ID during runtime; a misbehaved one will only shoot itself in the foot. (Note: I also suggest the IDs be chosen randomly to make them likely-unique.)

> A safer approach seems to be to arrange for the participating page on site B to be doing the capture itself. It controls its own layout, and can specify a crop region using its own `<div>` easily. Permissions policy would still ensure that the end user perceives the request as coming from site A, so that shouldn't be a problem.

I don't think the case of a single-frame application differs from a multi-frame application wrt to div-based-coordinates in any way other than that multi-frame applications need a crop-ID and single-frame applications can pass the div's DOM node as a parameter. I don't think ID-based referencing of divs is sufficiently more complicated than node-based referencing, and so I think we can go with the more general solution that also supports multi-frame-multi-origin applications.

**Comment by jan-ivar** [comment]
> multi-frame applications need a crop-ID and single-frame applications can pass the div's DOM node as a parameter. 

Why not have site L do the capture? Then there's no need to define this new crop-ID concept.

**Comment by eladalon1983** [comment]
> > multi-frame applications need a crop-ID and single-frame applications can pass the div's DOM node as a parameter.
> 
> Why not have site L do the capture? Then there's no need to define this new crop-ID concept.

Capturing is just the beginning of a frame's journey to remote users. Transmitting a video to a conference may involve a **significant** amount of application logic - especially with some upcoming/under-discussion features such as Breakout Box, Web Codecs, Insertable Streams, Lego Laser, etc. I don't think large applications like Slides/Docs, that normally do not handle any video-conferencing, should be expected to take on that responsibility. Letting a Slides-like application embed an iframe with a video-conferencing application (Meet/Zoom) is the natural thing, IMHO, and I think we should support such a use case.

**Comment by zenhack** [comment]
Alternative suggestion:

1. Make the media streams themselves [Transferable][1]
2. Provide an API to get a media stream from a DOM node

Then, rather than creating a full-tab stream and then cropping, then having to pass an ID, the application would just create a MediaStream for the DOM node in question, and pass it to the other application via postMessage() directly. This seems both simpler and more general.

In particular, making MediaStream (and possibly some other API objects) transferable would also solve some use cases I have for [Sandstorm][2], see discussion at https://github.com/w3c/webrtc-extensions/issues/64#issuecomment-764912264, whereas it's not clear the proposal here would.

[1]: https://developer.mozilla.org/en-US/docs/Web/API/Transferable
[2]: https://sandstorm.io

**Comment by eladalon1983** [comment]
@zenhack, this would be a very elegant solution if it is indeed possible to make media streams [Transferable](https://developer.mozilla.org/en-US/docs/Web/API/Transferable). But is that really the case? For example, how would you transfer a MediaStream [originating from a canvas](https://developer.mozilla.org/en-US/docs/Web/API/HTMLCanvasElement/captureStream) to a cross-origin iframe? What would that mean for the frames after the transfer? And assuming it's possible, would that not produce an attack vector between these two iframes?

**Comment by zenhack** [comment]
To make sure I understand your concern correctly: the issue is that `CanvasCaptureMediaStreamTrack` provides access to the underlying canvas object, and thus much of the other frame's DOM?

I agree it's probably best not to include that. Perhaps we want to limit this to functionality provided by the `MediaStreamTrack` interface (not subclasses in general). Spitballing: maybe we could add a new type `TransferableMediaStreamTrack`, whose constructor wraps another `MediaStreamTrack` and makes it transferable, but only exposes "safe" functionality? (In this case only the tracks would be transferable, not the `MediaStream` as a whole -- but I don't think that's terribly limiting).

**Comment by martinthomson** [comment]
@eladalon1983 though we haven't done the engineering in Gecko to enable transfer, at least we have the necessary markings.  If the stream that originates in a canvas is origin-clean, it would be marked as readable to the origin of the page.  If that page decides to transfer the content, we could allow the principal to be changed to that of the destination.  However, if the canvas becomes tainted, then the principal becomes the null origin, which would have to carry across also.

That's separate from having to do things like IPC for transferring between processes and all sorts of fiddly engineering like that.

Capturing DOM elements isn't exactly simple though: what about when cross-origin content is overlaid (using CSS `z-index`)?  If we read from the compositor, we *might* be able to get the un-occluded content.  It is tricky though.

**Comment by eladalon1983** [comment]
Making MediaStream transferable sounds like it could work well for mic/camera access (your use case is interesting, @zenhack). But I don't think it would work well in general. Take performance for example. Transferring 4k frames over IPCs would not be nice. Making MediaStreams transferrable is a much bigger scope project than what I'm suggesting. There are plenty of other implementation concerns which make me **suspect** Chrome would not be able to support this in the near future. (The general case, that is. Mic/camera - maybe?) I'd prefer continue pursuing our initially suggested API for the purpose of cropping at the moment.

> However, if the canvas becomes tainted, then the principal becomes the null origin, which would have to carry across also.

I think transferring a MediaStream is only interesting when it's done with its "contents" (the right term eludes me atm). For anything else, the recipient might as well just create their own MediaStream and configure it with everything else the original MediaStream had, since all other parts that I can think of are probably serializable and can be posted between frames already. Or am I overlooking something?

> Capturing DOM elements isn't exactly simple though: what about when cross-origin content is overlaid

I agree that capturing occluded content is risky, and does not mesh well with the kind of informed consent the user can give. It's for that reason that I am not suggesting DOM-element-capture, but rather I suggest cropping a capture of the entire tab to the coordinates specified by a div. With that approach, the pixels captured end up identical to those the local user is seeing. Cross-origin content is still an issue, but it's as much an issue before the crop as it is after. Wdyt?

**Comment by youennf** [comment]
I filed https://github.com/w3c/mediacapture-extensions/issues/16 so that we do not hijack this thread.

> here are plenty of other implementation concerns

Could you write them down in the above issue?

> Transferring 4k frames over IPCs would not be nice.

I think this is already happening for camera and screen capture in today browsers.
Also, compositing might happen out of process and the process doing the compositing is probably a good place to do tab capture.

In your specific usecase, the iframes might stay in the same process given they are on the same registrable domain.
Implementations could optimise that code path and skip IPC altogether.

I think it is important to look at perf and ergonomics.
I am wondering what your assessment of these approaches are in terms of ergonomics for the use case you have in mind and more generally.

**Comment by alvestrand** [comment]
The work on Insertable Streams and Breakout Box has brought the issues of transferring buffers between processes very much to the forefront.
The critical part is where things are exposed to Javascript - if we expose them to the user, they have to be in that user's process. As long as they don't need to be exposed, we can shuffle them around via references, which will often allow us to keep the "real" buffer in the process where they need to be and not move them out (for instance if they can remain in GPU buffers).

A MediaStreamTrack (MediaStreams are boring!) doesn't necessarily move much, but what making it transferable does is to move it to another context where the content *might* be exposed to Javascript - which means that we need code for every path that would allow such exposure. Which is, again, a nontrivial undertaking.

**Comment by eladalon1983** [comment]
> I filed [w3c/mediacapture-extensions#16](https://github.com/w3c/mediacapture-extensions/issues/16) so that we do not hijack this thread.

Thanks!

> > here are plenty of other implementation concerns
> 
> Could you write them down in the above issue?

Gladly.

> > Transferring 4k frames over IPCs would not be nice.
> 
> I think this is already happening for camera and screen capture in today browsers.

Yes, but adding yet more IPCs come at an added cost.

> Also, compositing might happen out of process and the process doing the compositing is probably a good place to do tab capture.
>
> In your specific usecase, the iframes might stay in the same process given they are on the same registrable domain.
> Implementations could optimise that code path and skip IPC altogether.

I agree that transferability can be simpler in some special cases, like mic/camera and possibly tab-capture. But AFAICT, not so in the general case. Are you suggesting to only allow transfer of MediaStreams between iframes that are same-site?

> I think it is important to look at perf and ergonomics.
> I am wondering what your assessment of these approaches are in terms of ergonomics for the use case you have in mind and more generally.

I completely agree that ergonomics in my particular use case would be superior with transferable streams. My problem is with the performance of the general case, and with the feasibility of implementing transferability only for a set of specific cases. I am also concerned that a solution that is ergonomic for a web-developer when it works (mic/camera, select special cases like tab capture), but sometimes fails for initially-non-obvious reasons (cross-origin MediaStream), might end up less ergonomic. Wdyt?

**Comment by youennf** [comment]
> I completely agree that ergonomics in my particular use case would be superior with transferable streams.

I would tend to give it a try. Prototyping it in Chromium over media capture insertable streams and transferable streams should give some insights.

> sometimes fails for initially-non-obvious reasons (cross-origin MediaStream), might end up less ergonomic. Wdyt?

If we specify it, we want to make it work for all cases the specification will support.
Some implementations might optimise the code path for only a reasonable subset.

**Comment by eladalon1983** [comment]
> > I completely agree that ergonomics in my particular use case would be superior with transferable streams.
> 
> I would tend to give it a try. Prototyping it in Chromium over media capture insertable streams and transferable streams should give some insights.

Sorry, I might have given the wrong impression with my emphatic assent. I agree that ergonomics for the application **would** be superior if this were developed, but I don't think that it is an easy thing to develop. I think Chrome is quite far from being able to support this.

**Comment by youennf** [comment]
> I think Chrome is quite far from being able to support this.

I haven't followed how much of media capture insertable streams is prototyped in Chrome.
If some Chrome version has support for MediaStreamTrackProcessor and MediaStreamTrackGenerator, what I thought could be tried is the following:
- In the iframe, call getDisplayMedia to get a track
- Use TrackProcessor to get the readableStream of video frames for that track
- Use postMessage to send the readableStream to the other iframe
- Use TrackGenerator to recreate a MediaStreamTrack from the transferred readableStream
- Use this track with RTCPeerConnection

**Comment by martinthomson** [comment]
To an earlier point regarding performance, if we were going to transfer media streams, it would be done in the same way we transfer between processes for rendering: shared memory.  The same applies to inline processing cases.  Of course, this only makes the engineering cost more, not less.

**Comment by zenhack** [comment]
Getting a track that maps to the actual pixels on the screen within the bounds of a node fits more with what I had in mind anyway.

I think being able to make a track based on a cropping and then transfer it would still probably be more ergonomic than having to make a crop_id and then transfer it, then have the other frame have to redeem the id for an actual object -- it's not clear to me what the indirection is buying you vs. having a certain type of track that is transferable, even if not all tracks are.

(Since my use case is different I don't want to hijack too much, but since a same-origin restriction was brought up I want to point out that I do actually need cross-origin transfers for what I'm doing).

**Comment by guidou** [comment]
If one were to transfer a MediaStreamTrack to an iframe with a different origin, wouldn't that make the transferred track a track with cross-origin content (from the perspective of the iframe)?
If that's the case, what should be done with such a track? mediacapture-main does not mention this possibility because it focuses on webcam and mic which can't be considered cross origin, but mediacapture-fromelement can and it says that "attempting to transfer the media using WebRTC results in no information being transmitted", and similar things for other APIs.
Chromium does not even support these "tainted" tracks. When a track becomes cross origin, it stops.

Wrt the possibility of prototyping this using Breakout Box, the transferability of VideoFrames in Chromium is limited to same-origin/same-process cases, so it's not possible to produce a useful shim using the approach @youennf proposes in https://github.com/w3c/mediacapture-screen-share/issues/158#issuecomment-765565070 . Full transferability to cross-origin (and potentially cross-process) requires more complex mechanisms such as @martinthomson mentions in https://github.com/w3c/mediacapture-screen-share/issues/158#issuecomment-765815053.

**Comment by jan-ivar** [comment]
> I filed w3c/mediacapture-extensions#16 so that we do not hijack this thread.

@youennf  Thanks, I've commented on transferring MediaStreamTrack there.

> Transmitting a video to a conference may involve a significant amount of application logic - especially with some upcoming/under-discussion features such as Breakout Box, Web Codecs, Insertable Streams, Lego Laser, etc.

Why, are you planning to blur the desktop? Or add overlays? Serious question, because, if you are, such app logic may belong more naturally in the top-level page, which actually owns, knows and controls what's being captured, than the sub-iframe.

Complexity arguments are double-edged swords: A complexity sufficient to scare away web developers from using browser features directly would count against standardizing such APIs in my book. I'd also keep in mind they're not standard (yet).

Demand for lower-level APIs presumably comes from developers who want this level of access to do work, and it's our job to make those APIs as simple and easy to use as possible. If sites decide to do this through JS libraries, hopefully, such libraries can easily be pulled into the top-level page as well, which is the one in charge of the merged experience.

> I don't think large applications like Slides/Docs, that normally do not handle any video-conferencing, should be expected to take on that responsibility.

I'd like to push back on this. I actually think they do. The original premise that a slides app beams itself into a web conference, suggested to me that the natural separation of concerns was for the slides app to own that part. This is the crux of better integrated HTML capture to me. As features grow, things like remote interaction with the content (remote mouse pointers etc.) all seem more naturally handled by the app that's capturing "itself".

As a corollary, if video-capture of DOM weren't available, then the "full web" approach to enable a remote HTML experience might be to recreate the DOM using existing web. That would definitely fall into the scope of the slides app, and would be significantly more work than video capture.

The alternative where one origin spies on another, seems like a part of `getDisplayMedia` we should refrain from building on, or at least not encourage.

A counter-argument perhaps might be needing to show a "self-view" of the capture, like WebEx does this, except where would one expect the self-view to appear? In slides or in the meet iframe?

As described, the main app here seems like slides, and whether we postMessage instructions or the results of said instructions seems like a detail.

Last, but not least, for this feature to be safe, it needs to be secured by something like COEP, as I propose in https://github.com/w3c/mediacapture-screen-share/issues/155. Those restrictions would need to apply to the top-level page, so we might as well assume it will need to be deeply involved.

**Comment by eladalon1983** [comment]
> Why, are you planning to blur the desktop? Or add overlays? Serious question, because, if you are, such app logic may belong more naturally in the top-level page

I foresee that video-conferencing applications will be using sophisticated upcoming WebRTC features, and some of these do not require any knowledge of the application or the video-content being captured from it. One example is the intention to allow applications to inject certain WebRTC modules. Another example is Web Codecs.

> A complexity sufficient to scare away web developers from using browser features

I argue that WebRTC is evolving to the point where power users - like applications **centered** around video-conferencing - can do much, but that this comes with added complexity when compared to less sophisticated WebRTC usage. Note that this complexity does **not** arise from the cropping-feature which I propose. The complexity I am referring to comes from the addition of Web Codecs, Breakout Box, module-injection, etc. - all of which are orthogonal to the proposed cropping feature. A non-conferencing application like Slides might wish to enjoy the benefits of all of these sophisticated features, despite not itself being a sophisticated WebRTC user.

True, enjoying power-features can be achieved through embedding a library. It can also be achieved by embedding a specialized application in an iframe. I don't think we should restrict applications to only one of these without good reason. IMHO, we don't have good reason here, as supporting both use cases (library/iframe) is easy.

> > I don't think large applications like Slides/Docs, that normally do not handle any video-conferencing, should be expected to take on that responsibility.
> 
> I'd like to push back on this. I actually think they do.

I'm in contact with Google Slides, and I know that they would prefer to keep conferencing logic in the Meet iframe. I think that browsers need to take into account the preferences of applications, and IMHO Google Slides' case is likely to be common to other applications as well. That is, the case of a large non-conferencing application wishing to embed a large conferencing-specific application in an iframe. Please note how little work would be needed to integrate these two applications if we use my twin suggestions of share-this-tab and crop-to-div(-by-id). Integration could then be accomplished with merely two messages posted between these frames - the video-conferencing iframe posts "I wish to start capturing content," and non-conferencing top-level application responds "the DIV you're looking for can be referenced through the crop-ID 'ca28ff965d'."

> Last, but not least, for this feature to be safe, it needs to be secured by something like COEP, as I propose in #155. Those restrictions would need to apply to the top-level page, so we might as well assume it will need to be deeply involved.

This is an orthogonal issue where we'll hopefully reach orthogonal agreement. I think that in the current discussion, it may safely be assumed that share-this-tab is indeed gated by your original COEP-based suggestion. (I mean, even if it ends up not being the case, we can discuss the interaction of cropping with whatever other gating we end up using once we know what that is.)

**Comment by eladalon1983** [comment]
Note [restrictOwnAudio](https://www.w3.org/TR/screen-capture/#dfn-restrictownaudio), btw. With this already-standardized constraint, a VC application that's capturing the entire tab can exclude audio that originates in the app's own document. So a PowerPoint-like app could embed a Meet-like VC app, and when the VC app captured the current tab's audio, it would easily be excluding remote participants' audio from the capture, but nothing else. This is another argument for supporting the use-case of an app living in one document cropping to the dimensions of a DIV living in another document.

**Comment by eladalon1983** [comment]
Background for those not involved - @jan-ivar, @youennf and I have been discussing whether a call to getViewportMedia should return the entire tab's viewport, or only of the viewport of the iframe from which getViewportMedia was called.

@jan-ivar and @youennf, here's another issue. When getViewportMedia is called, the user is prompted to approve the capture. It is very likely that the user agent will want to show the user a **visual preview** of what they are about to share. Showing anything short of the entire tab is going to be potentially misleading, since the iframe can always be resized at a later time to encompass a larger portion of the top-level. I think a reasonable user agent will always show a preview of the entire tab, even if the iframe - at the time gVM is invoked - is only p% of the tab.

I think we will have an inelegant and misleading API if we show the user[*] one thing, but hand the application something else. I think the elegant solution is to give the entire tab by default, and crop as a variant (additional parameter, later function call, etc.) Wdyt?

[*] The user can also be a web-app developer who will be quite confused after refactoring their code to call gVM from iframe/another-iframe/top-level, suddenly seeing the same preview but getting a differently cropped track, despite no explicit mention of cropping.

**Comment by eladalon1983** [comment]
I think there's a larger principle here - that good APIs are clear about what they do, and clarity is achieved by being explicit. If we phrase getViewportMedia as essentially "capture the entire tab's viewport," there is no question of what happens to occluding/occluded content - it's immediately clear.

However, if the API crops to the current iframe, then things become a lot less clear for the developer. Why are things cropped if no cropping was mentioned? And if we rebrand the API as getIframeMedia or something similar, things will be confusing in a different way - the difference from element-level capture would become unclear, with occluding/occluded content now behaving in unexpected ways.

IMHO, the clearest API is to just capture the entire tab, as getDisplayMedia does. Cropping can and should be an explicit extra.

**Comment by eladalon1983** [comment]
@jan-ivar and @youennf, could you PTAL? This is a blocking topic for [getViewportMedia](https://github.com/w3c/mediacapture-screen-share/issues/155).

**Comment by youennf** [comment]
To summarise:
- There is interest to capture a whole tab view, let's call it 'tab' option.
- There is interest to capture a document view, let's call it 'self' option.
- There is interest to capture a subpart of the document view, for instance an iframe of a document, maybe divs, let's call it 'node' option.

I think there is consensus for providing an option in the API to allow a web page to ask UA do the cropping.
It would be good to get confirmation if my understanding is correct.
In particular, cropping adds some complexity to the implementation to support this API in iframes for instance.

There is no consensus whether the default option value should be 'tab' or 'self'.

To make progress, I would go with adding an option to the API now.
It would take 'tab' or 'self' as values and we would make sure it could be extended to take nodes as a follow-up.

I would hope that the default option value becomes less crucial once the API offers both options.

**Comment by jan-ivar** [comment]
I would be open to an option if the default is "self".

I don't think we should add "node". Nodes don't all have viewports, and at that depth, I think it becomes surprising to users what is actually being captured (whatever is rendered at the absolute coordinates of that node, which based on layout can be entirely different nodes). The utility here seems low, and I'm not even sure I want to know the use cases.

While these same problems may exist at the frame level, I think people can handle and understand that. I see no need to compound the complexity.

**Comment by jan-ivar** [comment]
I don't think UAs are likely to offer preview. Users don't understand iframes, and the cropping is not a security guarantee (since the capture may contain stuff overlapping the iframe). So I see little value in a user agent communicating that the grant is of a section smaller than the page, even for "self" from an iframe.

**Comment by eladalon1983** [comment]
> I would be open to an option if the default is "self".

Or better yet - don't allow a default. Force the application to be explicit. Wdyt?
(Not a critical point for me, but otherwise we might get into a situation where initially, some browsers only supports the less technically-challenging non-default version, and nobody benefits from that.)

> So I see little value in a user agent communicating that the grant is of a section smaller than the page

That is precisely my point. I think user agents are very likely to provide a preview, as this gives a clear indication of what the user approves - much better than text that never gets read. The preview will **always** consist of the entire tab - independently cropping, because the cropped region can change unexpectedly (due to layout changes). Supporting (also?) an un-cropped variant of this API is therefore the elegant thing to do.

> it becomes surprising to users what is actually being captured

There is no room for surprise or confusion. The end-user consents to capturing the entire tab, always. Anything else is inscrutable to them. As far as the end-user knows, it's the application which crops the frames; the end-user neither knows nor cares that the UA serves as the scissors.

> I don't think we should add "node". .. The utility here seems low

I see cropping to an arbitrary iframe (not node) as a use-case of utmost importance. When capturing the current tab, it is very likely that you're capturing from a VC-iframe, but want to stream the content-iframe, where "content" is a doc/slide/game/video/you-name-it. Some have argued for making tracks/stream transferrable, but I don't see that as a good alternative, as I have explained.

**Comment by jan-ivar** [comment]
> Or better yet - don't allow a default. Force the application to be explicit. Wdyt?

That's redundant unless the caller is in an iframe. I'd want the default to be "self" if we go with an option.

> as this gives a clear indication of what the user approves

[Again](https://github.com/w3c/mediacapture-screen-share/issues/158#issuecomment-862889108), cropping is not a security guarantee, so no, they're not approving only the cropped area.

> (Not a critical point for me, but otherwise we might get into a situation where initially, some browsers only supports the less technically-challenging non-default version, and nobody benefits from that.)

([Crbug 2276](https://bugs.chromium.org/p/webrtc/issues/detail?id=2276) aside) how would one implement only a non-default value web-compatibly? 

> it becomes surprising to users what is actually being captured

Sorry I meant web-developer here, not user. My bad.

> I see cropping to an arbitrary iframe (not node) as a use-case of utmost importance.

Good we agree then on not adding "node".

**Comment by eladalon1983** [comment]
> Again, cropping is not a security guarantee, so no, they're not approving only the cropped area.

I beg you to read my messages more carefully. The user approves the entire tab. I've been arguing that very point all along. Like [here](https://github.com/w3c/mediacapture-screen-share/issues/158#issuecomment-863338821): "The end-user consents to capturing the entire tab, always." If you think this is not clear enough, I am always eager to hear how I can enhance the clarity of my writing.

> (Crbug 2276 aside) how would one implement only a non-default value web-compatibly?

By raising an exception, either compatibly or incompatibly. But that is not the best way forward. It would be better to align our visions: I do not consider cropping to the current iframe an important case, but I do consider cropping to an arbitrary iframe important. One is a subcase of the other. If we specify some support for cropping to an arbitrary iframe, then Chrome will implement that ability relatively soon. It will then be trivial to support "self" as well. That means no risk of a gap between specification and implementation; no compatibility issues; **no problem setting "self" as the default value**. A good compromise, I believe.

**Comment by youennf** [comment]
Another potential approach:
- Add an API for tab viewport capture, e.g. navigator.mediaDevices.getTabViewPortMedia()
- Add an API for document viewport capture, e.g. document.getViewPortMedia()
- Add an API for iframe viewport capture, e.g. HTMLIFrameElement.getViewPortMedia()
All these methods would call the same underlying algorithm which would take a viewport as internal parameter.
This underlying algorithm would be responsible of permission policy, prompting, creation of MediaStreamTrack...

With this approach, there is no default option to select in the spec and no surprise from web developers on what they will get.
Another advantage is easier discovery of whether tab/self/iframe capture are supported: browsers can gradually implement them and applications gracefully fallback.
Also, if we want to extend to other nodes (all elements for instance), getViewPortMedia() can be defined for those nodes.

**Comment by eladalon1983** [comment]
> Another potential approach:
> 
> * Add an API for tab viewport capture, e.g. navigator.mediaDevices.getTabViewPortMedia()
> * Add an API for document viewport capture, e.g. document.getViewPortMedia()
> * Add an API for iframe viewport capture, e.g. HTMLIFrameElement.getViewPortMedia()
>   All these methods would call the same underlying algorithm which would take a viewport as internal parameter.
>   This underlying algorithm would be responsible of permission policy, prompting, creation of MediaStreamTrack...
> 
> With this approach, there is no default option to select in the spec and no surprise from web developers on what they will get.
> Another advantage is easier discovery of whether tab/self/iframe capture are supported: browsers can gradually implement them and applications gracefully fallback.
> Also, if we want to extend to other nodes (all elements for instance), getViewPortMedia() can be defined for those nodes.

I'd be happy with this approach.
